parlai/agents/bert_ranker/helpers.py

```diff
@@ -0,0 +1,159 @@
+        # deduce bert output dim from the size of the pooler in bert
+        bert_output_dim = bert_model.pooler.dense.weight.size()[0]
+
+        if add_transformer_layer:
+            config_for_one_layer = BertConfig(
+                0, hidden_size=bert_output_dim, num_attention_heads=int(
+                    bert_output_dim / 64), intermediate_size=3072, hidden_act="gelu")
+            self.additional_transformer_layer = BertLayer(config_for_one_layer)
+        self.additional_linear_layer = torch.nn.Linear(bert_output_dim, output_dim)
+        self.bert_model = bert_model
+
+    def forward(self, token_ids, segment_ids, attention_mask):
+        output_bert, _ = self.bert_model(token_ids, segment_ids, attention_mask)
+        # output_bert is a list of 12 (for bert base) layers.
+        layer_of_interest = output_bert[self.layer_pulled]
+        if self.add_transformer_layer:
+            # Follow up by yet another transformer layer
+            extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
+            extended_attention_mask = extended_attention_mask.to(
+                dtype=next(self.parameters()).dtype)  # fp16 compatibility
+            extended_attention_mask = (1.0 - extended_attention_mask) * -10000.0
+            embeddings = self.additional_transformer_layer(
+                layer_of_interest, extended_attention_mask)[:, 0, :]
+        else:
+            embeddings = layer_of_interest[:, 0, :]
+        return self.additional_linear_layer(embeddings)
+
+
+def surround(idx_vector, start_idx, end_idx):
+    """ Surround the vector by start_idx and end_idx
+    """
+    start_tensor = idx_vector.new_tensor([start_idx])
+    end_tensor = idx_vector.new_tensor([end_idx])
+    return torch.cat([start_tensor, idx_vector, end_tensor], 0)
+
+
+patterns_optimizer = {
+    "additional_layers": ["additional"],
+    "top_layer": [
+        "additional",
+        "bert_model.encoder.layer.11."],
+    "top4_layers": [
+        "additional",
+        "bert_model.encoder.layer.11.",
+        "encoder.layer.10.",
+        "encoder.layer.9.",
+        "encoder.layer.8"],
+    "all_encoder_layers": [
+        "additional",
+        "bert_model.encoder.layer"],
+    "all": [
+        "additional",
+        "bert_model.encoder.layer",
+        "bert_model.embeddings"],
+}
+
+
+def get_bert_optimizer(models, type_optimization, number_iterations, proportion_warmup,
+                       learning_rate):
+    """
+        Provides an optimizer already configured with weight decay for the parameters
+        that need it. Weight decay is left standard at 0.01
+    """
+    if type_optimization not in patterns_optimizer:
+        print("Error. Type optimizer must be one of %s" %
+              (str(patterns_optimizer.keys())))
+    parameters_with_decay = []
+    parameters_with_decay_names = []
+    parameters_without_decay = []
+    parameters_without_decay_names = []
+    no_decay = ['bias', 'gamma', 'beta']
```

Comment:
is that from the paper, that these aren't decayed?

Reply:
Seems to be from their code
https://github.com/google-research/bert/blob/master/optimization.py#L65